compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/inference/FirBuilderInferenceSession.kt

```diff
@@ -203,9 +201,24 @@ class FirBuilderInferenceSession(
         val commonSystemSubstitutor = commonSystem.buildCurrentSubstitutor() as ConeSubstitutor
         val nonFixedTypesToResultSubstitutor = ConeComposedSubstitutor(commonSystemSubstitutor, nonFixedToVariablesSubstitutor)
         val completionResultsWriter = components.callCompleter.createCompletionResultsWriter(nonFixedTypesToResultSubstitutor)
+
+        for ((completedCall, candidate) in commonCalls) {
+            // TODO: Only update return type? Should we need to visit all appearances of unsubstituted postponed variables in types?
```

Comment:
Now I found a concrete example that shows more places of remaining stub types: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/codegen/box/builderInference/lackOfNullCheckOnNullableInsideBuild.kt#L25

The corresponding IR text looks like this:
```
                  CALL 'public final fun resolve (resolver: kotlin.Function1<<root>.Foo, T of <root>.Builder>): kotlin.Unit declared in <root>.Builder' type=kotlin.Unit origin=null
                    $this: GET_VAR '<this>: <root>.Builder<kotlin.String?> declared in <root>.LackOfNullCheckOnNullableInsideBuildKt.box.<anonymous>' type=<root>.Builder<IrErrorType> origin=null
                    resolver: BLOCK type=kotlin.Function1<<root>.Foo, IrErrorType> origin=LAMBDA
                      FUN LOCAL_FUNCTION_FOR_LAMBDA name:<anonymous> visibility:local modality:FINAL <> (it:<root>.Foo) returnType:IrErrorType
                        VALUE_PARAMETER name:it index:0 type:<root>.Foo
                        BLOCK_BODY
                          RETURN type=kotlin.Nothing from='local final fun <anonymous> (it: <root>.Foo): IrErrorType declared in <root>.LackOfNullCheckOnNullableInsideBuildKt.box.<anonymous>'
                            CALL 'public final fun <get-string> (): kotlin.String? declared in <root>.Foo' type=kotlin.String? origin=GET_PROPERTY
                              $this: GET_VAR 'it: <root>.Foo declared in <root>.LackOfNullCheckOnNullableInsideBuildKt.box.<anonymous>.<anonymous>' type=<root>.Foo origin=null
                      FUNCTION_REFERENCE 'local final fun <anonymous> (it: <root>.Foo): IrErrorType declared in <root>.LackOfNullCheckOnNullableInsideBuildKt.box.<anonymous>' type=kotlin.Function1<<root>.Foo, IrErrorType> origin=LAMBDA reflectionTarget=null
```
See 6 `IrErrorType` appearances, which are due to remaining stub type. By using the same substitutor, I can see that, e.g., dispatch receiver expression's type can be rescued easily.

But, still, that---dealing with type ref one-by-one---seems not scalable. I can see, e.g., `FirTypeResolver(Impl)?` and `Fir(Specific)?TypeResolveTransformer`. Does it make sense to introduce, say, `FirStubTypeSubstitutor` around this `FirBuilderInferenceSession`, which will be dedicated to visit only type references with stub types and substitute it with proper final type according to the composed substitutor here?

Reply:
I was convinced that we eventually need something similar, so I tried it. See my latest/5th commit, which can unmute some of builder inference bb tests.